### PR TITLE
Creating a separate module for StackHCIVM

### PR DESCRIPTION
**Pull request by hvedati**
## Checklist

- [ ] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Comment by VeryEarly** [comment]
please adjust folder structure to "src/StackHCIVm/StackHCIVm.Autorest"

File: src/StackHCIVm/README.md [hunk 1]
```diff
@@ -0,0 +1,646 @@
...
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/galleryImages.json 
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/marketplaceGalleryImages.json
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/networkInterfaces.json
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/storageContainers.json
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/virtualHardDisks.json
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/virtualMachineInstances.json
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/logicalNetworks.json
+  - $(repo)/specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-03-15-preview/HybridCompute.json
+
+module-version: 1.1.0
```
Comment (VeryEarly) [inline src/StackHCIVm/README.md:72]: 0.1.0

File: src/StackHCIVm/README.md [hunk 2]
```diff
@@ -0,0 +1,646 @@
...
+
+### AutoRest Configuration
+
+> see https://aka.ms/autorest
+
+```yaml
+require:
+  - $(this-folder)/../readme.azure.noprofile.md
+repo: azure-rest-api-specs
+# lock the commit
```
Comment (VeryEarly) [inline src/StackHCIVm/README.md:60]: missed line of commit

**Review by isra-fel — CHANGES_REQUESTED** [review]
Regarding the casing of the module, we have a convention that if an abbreviation is longer than 2 (> 2) we use Pascal case, for example, `Restart-AzVmss`, `New-AzSqlDatabase`. But if the length is less or equal to two (<= 2) we use UPPERCASE, for example, `New-AzVM`, `Remove-AzCosmosDBTable`.
To follow the convention, the module would be named `Az.StackHCIVM`. Please also make it consistent in the cmdlets, source file names, docs, etc.

**Comment by NoriZC** [comment]
Hi @hvedati,
As you are separating StackHCIVM from StackHCI, we will think this module as a new one.

For new modules onboard, we recommend using autorest powershell v4 for code generation. But in your module, you are still using  the namespace which comes from the v3 generated module. For current case, you shoud update the namespaces in all custom files: `Microsoft.Azure.PowerShell.Cmdlets.StackHCIVM.Models.Api20230901Preview.IVirtualMachineInstance` => Microsoft.Azure.PowerShell.Cmdlets.StackHCIVM.Models.IVirtualMachineInstance`

The reason why tou are not reproducing the error locally is you are not using the newest version of autorest locally.

Please do following steps:
- Update all the custom files to fit the autorest powershell v4 generated namespaces.
- You should use directive to update the module name from StackHCIVm to StackHCIVM

File: src/StackHCIVm/StackHCIVm.Autorest/README.md [hunk 3]
```diff
@@ -0,0 +1,647 @@
...
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/networkInterfaces.json
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/storageContainers.json
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/virtualHardDisks.json
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/virtualMachineInstances.json
+  - $(repo)/specification/azurestackhci/resource-manager/Microsoft.AzureStackHCI/preview/2023-09-01-preview/logicalNetworks.json
+  - $(repo)/specification/hybridcompute/resource-manager/Microsoft.HybridCompute/preview/2023-03-15-preview/HybridCompute.json
+
+module-version: 0.1.0
+title: StackHCIVM
+service-name: StackHCIVM
```
Comment (wyunchi-ms) [inline src/StackHCIVm/StackHCIVm.Autorest/README.md:75]: Please replace all the `StackHCIVM` with `StackHCIVm` in this project

**Review by NoriZC — APPROVED** [review]

**Review by isra-fel — CHANGES_REQUESTED** [review]
Hi @hvedati

1. Turned out there are a few misspellings left. Please check the screenshot.
2. Custom cmdlets need to include some common parameters to support testing and proxy, etc. Please follow [this guide](https://eng.ms/docs/cloud-ai-platform/azure-core/azure-management-and-platforms/control-plane-bburns/azure-cli-tools-azure-cli-powershell-and-terraform/azure-cli-tools/onboarding/azurepowershell/custom_code_test_coverage#forget-to-add-common-parameters-for-custom-code) to add them.
3. After adding the common parameters, the custom cmdlets can be recorded normally. Please make sure to add test cases.



image

**Review by NoriZC — APPROVED** [review]

**Pull request merged**